elk/utils/multi_gpu.py

```diff
@@ -0,0 +1,184 @@
+from contextlib import nullcontext, redirect_stdout
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, Type
+
+import torch
+from accelerate import infer_auto_device_map, init_empty_weights
+from torch import dtype
+from torch.nn import Module
+from transformers import PreTrainedModel
+
+from elk.utils import instantiate_model, select_usable_devices
+from elk.utils.gpu_utils import get_available_memory_for_devices
+
+if TYPE_CHECKING:
+    from elk import Extract
+
+
+@dataclass
+class ModelDevices:
+    # The devices to instantiate a single model on
+    first_device: str
+    other_devices: list[str]
+
+    @property
+    def is_single_gpu(self) -> bool:
+        return len(self.other_devices) == 0
+
+    @property
+    def used_devices(self) -> list[str]:
+        return [self.first_device] + self.other_devices
+
+
+def instantiate_model_with_devices(
+    cfg: "Extract", device_config: ModelDevices, is_verbose: bool, **kwargs
+) -> PreTrainedModel:
+    first_device = device_config.first_device
+    if cfg.int8:
+        # Required by `bitsandbytes`
+        torch_dtype = torch.float16
+    elif device_config == "cpu":
+        torch_dtype = torch.float32
+    else:
+        torch_dtype = "auto"
+
+    # TODO: Maybe we should ensure the device map is the same
+    # for all the extract processes? This is because the device map
+    # can affect performance highly and its annoying if one process
+    # is using a different device map than the others.
+    device_map = (
+        {"": first_device}
+        if device_config.is_single_gpu
+        else create_device_map(
+            model_str=cfg.model,
+            use_8bit=cfg.int8,
+            torch_dtype=torch_dtype,
+            model_devices=device_config,
+            verbose=is_verbose,
+        )
+    )
```

Comment:
the device map can be different if e.g., gpus 1-7 are all empty. But gpu 0 happens to have someone else's 5 gb model there. 
Suppose you have 4 workers (2 gpus each)

The first worker will have a different device map from the rest, since it'll take into account the 5gb model memory being used.

The device map can sometimes heavily affect performance. In my experience when u can unlucky like (1.5x slower?). This is because you maybe you'll split the model in a suboptimal spot. Like exactly where the model is going to send much more tensors between the layers.

So ideally all the workers would use the same device map so you can approximate the time take to process the dataset better.